sound/soc/sof/sof-priv.h

```diff
@@ -365,6 +365,25 @@ struct sof_ipc_pm_ops {
 	int (*ctx_restore)(struct snd_sof_dev *sdev);
 };
 
+/**
+ * struct sof_ipc_fw_loader_ops - IPC/FW-specific loader ops
+ * @validate:		Function pointer for validating the firmware image
+ * @parse_ext_manifest:	Function pointer for parsing the manifest of the firmware
+ * @load_fw_to_dsp:	Optional function pointer for loading the firmware to the
+ *			DSP.
+ *			The function implements generic, hardware independent way
+ *			of loading the initial firmware and its modules (if any).
+ * @query_fw_configuration: Optional function pointer to query information and
+ *			configuration from the booted firmware.
+ *			Executed after the first succesfull firmware boot.
+ */
+struct sof_ipc_fw_loader_ops {
+	int (*validate)(struct snd_sof_dev *sdev);
+	size_t (*parse_ext_manifest)(struct snd_sof_dev *sdev);
+	int (*load_fw_to_dsp)(struct snd_sof_dev *sdev);
+	void (*query_fw_configuration)(struct snd_sof_dev *sdev);
```

Comment:
@bardliao,
```c
* @query_fw_configuration: Optional function pointer to query information and
 *			configuration from the booted firmware.
 *			Executed after the first succesfull firmware boot.
```
The idea behind this is that IPC3 sends configuration info along with the FW_READY message, so we can get it right at that point.
With IPC4 however we will receive a simple message informing us that the firmware has booted up and we need to _ask_ for the configuration via a different request (get large data).

Basically this will allow us in the future to support both type of firmware boot completion.

The `query_fw_configuration` is to use IPC to get the information from the booted firmware and not from the raw image itself. It is kind of equivalent to a deferred fw_ready in IPC3 (we read the configuration from the mailbox in IPC3).

Reply:
@bardliao, see the last patch on when this is going to be executed. The problem is that we can not send messages from the rx path directly, so we need to make sure that sequencing is correct.

Reply:
@ujfalusi Will the configuration be stored somewhere in sdev or will we handle the configuration inside  `query_fw_configuration`?

Reply:
@bardliao, It depends, I guess. Some information can be used within the callback adjust sdev config options, do some checks, print debug information, some can be saved for later use.

In IPC3 we would discard the fw_ready information received from the firmware and use the info derived from the manifest part of the firmware file (we do a memcmp() to make sure they are identical).

I think this is not possible for IPC4.